libmount/src/context.c

```diff
@@ -2577,6 +2583,117 @@ int mnt_context_wait_for_children(struct libmnt_context *cxt,
 	return 0;
 }
 
+/**
+ * mnt_context_set_target_ns:
+ * @cxt: mount context
+ * @path: path to target namespace or NULL
+ *
+ * Sets target namespace to namespace represented by @path. If @path is NULL,
+ * target namespace is cleared.
+ *
+ * Returns: 0 on success, negative number in case of error.
+ */
+int mnt_context_set_target_ns(struct libmnt_context *cxt, const char *path)
+{
+	int rc = 0;
+
+	int tmp;
+	if (!cxt)
+		return -EINVAL;
+
+	DBG(CXT, ul_debugobj(cxt, "Setting %s as target namespace", path));
+
+	if (!path) {
+		close(cxt->namespaces[MNT_NS_ORIG].fd);
+		cxt->namespaces[MNT_NS_ORIG].fd = -1;
+		if (cxt->namespaces[MNT_NS_TGT].fd != -1) {
+			close(cxt->namespaces[MNT_NS_TGT].fd);
+			cxt->namespaces[MNT_NS_TGT].fd = -1;
+			mnt_unref_cache(cxt->namespaces[MNT_NS_TGT].cache);
+		}
+		return 0;
+	}
+
+	tmp = open(path, O_RDONLY | O_CLOEXEC);
+	if (tmp == -1)
+		rc = -errno;
+
+	if (!rc && cxt->namespaces[MNT_NS_ORIG].fd == -1) {
+		cxt->namespaces[MNT_NS_ORIG].fd = open("/proc/self/ns/mnt", O_RDONLY | O_CLOEXEC);
+		if (cxt->namespaces[MNT_NS_ORIG].fd == -1)
+			rc = -errno;
+	}
+
+	/* test whether namespace switching works */
+	if (!rc) {
+		DBG(CXT, ul_debugobj(cxt, "Trying whether namespace is valid"));
+		if (setns(tmp, CLONE_NEWNS)
+		    || setns(cxt->namespaces[MNT_NS_ORIG].fd, CLONE_NEWNS)) {
+			rc = -errno;
+			DBG(CXT, ul_debugobj(cxt, "setns(2) failed [errno=%d %m]", errno));
+		}
+	}
+	if (rc)
+		goto err;
+
+	if (cxt->namespaces[MNT_NS_TGT].fd != -1) {
+		mnt_unref_cache(cxt->namespaces[MNT_NS_TGT].cache);
+		close(cxt->namespaces[MNT_NS_TGT].fd);
+	}
+
+	cxt->namespaces[MNT_NS_TGT].fd = tmp;
+	cxt->namespaces[MNT_NS_TGT].cache = NULL;
+
+	return 0;
+
+err:
+	/* revert to original state */
+	close(tmp);
+	if (cxt->namespaces[MNT_NS_TGT].fd == -1) {
+		close(cxt->namespaces[MNT_NS_ORIG].fd);
+		cxt->namespaces[MNT_NS_ORIG].fd = -1;
+	}
+	return rc;
+}
+
+/*
+ * Switch to namespace specified by ns_idx
+ * Returns either index of previous namespace or -errno
+ */
+int mnt_context_switch_ns(struct libmnt_context *cxt, int ns_idx)
```

Comment:
If we want to support this in helpers than we probably need to export it by library API, because it's possible that a helper need any NS sensitive operations too.

This is why I think it would be better to introduce more functions and "struct libmnt_ns" (see comment below) and more hide implementation details.